File: FastDrawingVisual.RenderingCore/Backends/D3D9/D3D9Backend.cs
extern alias D3D9Proxy;

using FastDrawingVisual.CommandRuntime;
using System;
using System.Windows;
using System.Windows.Interop;
using Proxy = D3D9Proxy::FastDrawingVisual.NativeProxy.NativeProxy;

namespace FastDrawingVisual.Rendering.Backends
{
    public sealed class D3D9Backend : IRenderBackend, ILayeredFrameSink, ID3D9SurfaceProvider, ID3D9PresentationController, IRenderBackendReadiness
    {
        private IntPtr _nativeRenderer;
        private HwndSource? _fallbackHwndSource;
        private int _width;
        private int _height;
        private bool _isInitialized;
        private bool _isDeviceLost;
        private bool _isDisposed;
        private bool _lastReadyState;

        public bool IsReadyForRendering => !_isDisposed && _isInitialized && !_isDeviceLost && _nativeRenderer != IntPtr.Zero;

        public event Action? ReadyStateChanged;

        public bool Initialize(int width, int height)
        {
            ThrowIfDisposed();
            if (width <= 0 || height <= 0)
                throw new ArgumentException("Width and height must be greater than zero.");

            _width = width;
            _height = height;

            if (_nativeRenderer == IntPtr.Zero)
            {
                _nativeRenderer = CreateNativeRenderer(width, height);
                if (_nativeRenderer == IntPtr.Zero)
                {
                    UpdateReadyState();
                    return false;
                }
            }

            _isInitialized = true;
            _isDeviceLost = false;
            UpdateReadyState();
            return true;
        }

        public void Resize(int width, int height)
        {
            ThrowIfDisposed();
            if (width <= 0 || height <= 0)
                throw new ArgumentException("Width and height must be greater than zero.");

            _width = width;
            _height = height;

            if (_nativeRenderer == IntPtr.Zero || !_isInitialized || _isDeviceLost)
            {
                UpdateReadyState();
                return;
            }

            SafeResizeNative(width, height);
            UpdateReadyState();
        }

        public void SubmitFrame(in BridgeLayeredFramePacket frame)
        {
            ThrowIfDisposed();
            if (!IsReadyForRendering || !frame.HasAnyCommands)
                return;

            try
            {
                for (int layerIndex = 0; layerIndex < BridgeLayeredFramePacket.MaxLayerCount; layerIndex++)
                {
                    var layer = frame.GetLayer(layerIndex);
                    if (!layer.HasCommands)
                        continue;

                    Proxy.SubmitCommands(
                        _nativeRenderer,
                        layer.CommandPointer,
                        layer.CommandBytes,
                        layer.BlobPointer,
                        layer.BlobBytes);
                }
            }
            catch
            {
                _isDeviceLost = true;
                UpdateReadyState();
            }
        }

        public IntPtr GetSurface9()
        {
            ThrowIfDisposed();
            if (!IsReadyForRendering)
                return IntPtr.Zero;

            try
            {
                var surface = IntPtr.Zero;
                return Proxy.TryAcquirePresentSurface(_nativeRenderer, ref surface) ? surface : IntPtr.Zero;
            }
            catch
            {
                _isDeviceLost = true;
                UpdateReadyState();
                return IntPtr.Zero;
            }
        }

        public bool CopyReadyToPresentSurface()
        {
            ThrowIfDisposed();
            if (!IsReadyForRendering)
                return false;

            try
            {
                return Proxy.CopyReadyToPresentSurface(_nativeRenderer);
            }
            catch
            {
                _isDeviceLost = true;
                UpdateReadyState();
                return false;
            }
        }

        public void NotifyFrontBufferAvailable(bool available)
        {
            ThrowIfDisposed();

            if (!available)
            {
                _isInitialized = false;
                _isDeviceLost = true;
                SafeNotifyFrontBufferAvailable(false);
                UpdateReadyState();
                return;
            }

            if (_width <= 0 || _height <= 0)
                return;

            if (_nativeRenderer == IntPtr.Zero)
            {
                _nativeRenderer = CreateNativeRenderer(_width, _height);
                if (_nativeRenderer == IntPtr.Zero)
                {
                    UpdateReadyState();
                    return;
                }
            }

            SafeNotifyFrontBufferAvailable(true);
            if (!SafeResizeNative(_width, _height))
            {
                UpdateReadyState();
                return;
            }

            _isDeviceLost = false;
            _isInitialized = true;
            UpdateReadyState();
        }

        public bool TryGetCapability<TCapability>(out TCapability? capability)
            where TCapability : class
        {
            capability = this as TCapability;
            return capability != null;
        }

        public void Dispose()
        {
            if (_isDisposed)
                return;

            _isDisposed = true;
            DestroyNativeRenderer();
            _fallbackHwndSource?.Dispose();
            _fallbackHwndSource = null;
            UpdateReadyState();
        }

        private IntPtr CreateNativeRenderer(int width, int height)
        {
            var hwnd = GetOrCreateDeviceHwnd();
            if (hwnd == IntPtr.Zero)
                return IntPtr.Zero;

            try
            {
                return Proxy.CreateRenderer(hwnd, width, height);
            }
            catch
            {
                return IntPtr.Zero;
            }
        }

        private bool SafeResizeNative(int width, int height)
        {
            if (_nativeRenderer == IntPtr.Zero)
                return false;

            try
            {
                if (!Proxy.Resize(_nativeRenderer, width, height))
                {
                    _isDeviceLost = true;
                    return false;
                }

                return true;
            }
            catch
            {
                _isDeviceLost = true;
                return false;
            }
        }

        private void SafeNotifyFrontBufferAvailable(bool available)
        {
            if (_nativeRenderer == IntPtr.Zero)
                return;

            try
            {
                Proxy.OnFrontBufferAvailable(_nativeRenderer, available);
            }
            catch
            {
                _isDeviceLost = true;
            }
        }

        private void DestroyNativeRenderer()
        {
            if (_nativeRenderer == IntPtr.Zero)
                return;

            try
            {
                Proxy.DestroyRenderer(_nativeRenderer);
            }
            catch
            {
            }
            finally
            {
                _nativeRenderer = IntPtr.Zero;
                _isInitialized = false;
                _isDeviceLost = false;
            }
        }

        private IntPtr GetOrCreateDeviceHwnd()
        {
            var mainWindow = Application.Current?.MainWindow;
            if (mainWindow != null)
            {
                var mainHwnd = new WindowInteropHelper(mainWindow).Handle;
                if (mainHwnd != IntPtr.Zero)
                    return mainHwnd;
            }

            if (_fallbackHwndSource == null)
            {
                var parameters = new HwndSourceParameters("FastDrawingVisual.D3D9Backend")
                {
                    Width = 1,
                    Height = 1,
                    WindowStyle = unchecked((int)0x80000000),
                };
                _fallbackHwndSource = new HwndSource(parameters);
            }

            return _fallbackHwndSource.Handle;
        }

        private void UpdateReadyState()
        {
            var isReady = IsReadyForRendering;
            if (_lastReadyState == isReady)
                return;

            _lastReadyState = isReady;
            ReadyStateChanged?.Invoke();
        }

        private void ThrowIfDisposed()
        {
            if (_isDisposed)
                throw new ObjectDisposedException(nameof(D3D9Backend));
        }
    }
}

File: FastDrawingVisual.NativeProxy.D3D9/D3D9Backend.cpp
#include "D3D9Backend.h"

#include <new>
#include <vcclr.h>

using namespace System::Windows;
using namespace System::Windows::Interop;

namespace FastDrawingVisual::Rendering::Backends {

D3D9Backend::D3D9Backend() {}

D3D9Backend::~D3D9Backend() {
  if (isDisposed_) {
    return;
  }

  DestroyRenderer();
  if (fallbackHwndSource_ != nullptr) {
    delete fallbackHwndSource_;
    fallbackHwndSource_ = nullptr;
  }

  isDisposed_ = true;
  UpdateReadyState();
}

D3D9Backend::!D3D9Backend() {
  DestroyRenderer();
  fallbackHwndSource_ = nullptr;
  isDisposed_ = true;
}

bool D3D9Backend::Initialize(int width, int height) {
  ThrowIfDisposed();
  if (width <= 0 || height <= 0) {
    throw gcnew ArgumentException(
        "Width and height must be greater than zero.");
  }

  width_ = width;
  height_ = height;

  if (renderer_ == nullptr && !CreateNativeRenderer(width, height)) {
    UpdateReadyState();
    return false;
  }

  isInitialized_ = true;
  isFaulted_ = false;
  UpdateReadyState();
  return true;
}

void D3D9Backend::Resize(int width, int height) {
  ThrowIfDisposed();
  if (width <= 0 || height <= 0) {
    throw gcnew ArgumentException(
        "Width and height must be greater than zero.");
  }

  width_ = width;
  height_ = height;

  if (renderer_ == nullptr || !isInitialized_ || isFaulted_) {
    UpdateReadyState();
    return;
  }

  if (FAILED(renderer_->Resize(width, height))) {
    isFaulted_ = true;
  }

  UpdateReadyState();
}

IDrawingContext ^D3D9Backend::CreateDrawingContext(int width, int height) {
  ThrowIfDisposed();
  if (!IsReadyForRendering) {
    return nullptr;
  }

  return gcnew LayeredCommandRecordingContext(
      width, height,
      gcnew Action<LayeredFramePacket>(this, &D3D9Backend::SubmitFrame));
}

IntPtr D3D9Backend::GetSurface9() {
  ThrowIfDisposed();
  if (!IsReadyForRendering) {
    return IntPtr::Zero;
  }

  void *surface = nullptr;
  const HRESULT hr =
      renderer_ == nullptr ? E_UNEXPECTED : renderer_->TryAcquirePresentSurface(&surface);
  if (FAILED(hr)) {
    isFaulted_ = true;
    UpdateReadyState();
    return IntPtr::Zero;
  }

  return IntPtr(surface);
}

bool D3D9Backend::CopyReadyToPresentSurface() {
  ThrowIfDisposed();
  if (!IsReadyForRendering) {
    return false;
  }

  const HRESULT hr =
      renderer_ == nullptr ? E_UNEXPECTED : renderer_->CopyReadyToPresentSurface();
  if (FAILED(hr)) {
    isFaulted_ = true;
    UpdateReadyState();
    return false;
  }

  return hr == S_OK;
}

void D3D9Backend::NotifyFrontBufferAvailable(bool available) {
  ThrowIfDisposed();

  if (!available) {
    isInitialized_ = false;
    isFaulted_ = true;
    if (renderer_ != nullptr) {
      renderer_->OnFrontBufferAvailable(false);
    }

    UpdateReadyState();
    return;
  }

  if (width_ <= 0 || height_ <= 0) {
    return;
  }

  if (renderer_ == nullptr && !CreateNativeRenderer(width_, height_)) {
    UpdateReadyState();
    return;
  }

  if (renderer_ != nullptr) {
    renderer_->OnFrontBufferAvailable(true);
    if (FAILED(renderer_->Resize(width_, height_))) {
      isFaulted_ = true;
      UpdateReadyState();
      return;
    }
  }

  isFaulted_ = false;
  isInitialized_ = true;
  UpdateReadyState();
}

bool D3D9Backend::CreateNativeRenderer(int width, int height) {
  IntPtr hwnd = GetOrCreateDeviceHwnd();
  if (hwnd == IntPtr::Zero) {
    return false;
  }

  auto *renderer = new (std::nothrow)
      fdv::d3d9::D3D9Renderer(static_cast<HWND>(hwnd.ToPointer()), width, height);
  if (renderer == nullptr) {
    return false;
  }

  const HRESULT initializeHr = renderer->Initialize();
  if (FAILED(initializeHr)) {
    delete renderer;
    return false;
  }

  renderer->RegisterMetrics();
  renderer_ = renderer;
  return true;
}

void D3D9Backend::SubmitFrame(LayeredFramePacket frame) {
  ThrowIfDisposed();
  if (!IsReadyForRendering || !frame.HasAnyCommands) {
    return;
  }

  LayeredFramePacket frameCopy = frame;
  pin_ptr<LayeredFramePacket> framePacket = &frameCopy;

  if (renderer_ == nullptr ||
      FAILED(renderer_->SubmitLayeredCommands(framePacket,
                                              sizeof(LayeredFramePacket)))) {
    isFaulted_ = true;
  }

  UpdateReadyState();
}

void D3D9Backend::DestroyRenderer() {
  if (renderer_ == nullptr) {
    return;
  }

  renderer_->UnregisterMetrics();
  delete renderer_;
  renderer_ = nullptr;
  isInitialized_ = false;
  isFaulted_ = false;
}

IntPtr D3D9Backend::GetOrCreateDeviceHwnd() {
  Application^ app = Application::Current;
  Window^ mainWindow = app != nullptr ? app->MainWindow : nullptr;
  if (mainWindow != nullptr) {
    WindowInteropHelper helper(mainWindow);
    IntPtr mainHwnd = helper.Handle;
    if (mainHwnd != IntPtr::Zero) {
      return mainHwnd;
    }
  }

  if (fallbackHwndSource_ == nullptr) {
    HwndSourceParameters parameters(gcnew String(L"FastDrawingVisual.D3D9Backend"));
    parameters.Width = 1;
    parameters.Height = 1;
    parameters.WindowStyle = static_cast<int>(0x80000000u);
    fallbackHwndSource_ = gcnew HwndSource(parameters);
  }

  return fallbackHwndSource_->Handle;
}

void D3D9Backend::UpdateReadyState() {
  const bool isReady = IsReadyForRendering;
  if (lastReadyState_ == isReady) {
    return;
  }

  lastReadyState_ = isReady;
  if (readyStateChanged_ != nullptr) {
    readyStateChanged_();
  }
}

void D3D9Backend::ReadyStateChanged::add(Action^ handler) {
  readyStateChanged_ += handler;
}

void D3D9Backend::ReadyStateChanged::remove(Action^ handler) {
  readyStateChanged_ -= handler;
}

void D3D9Backend::ThrowIfDisposed() {
  if (isDisposed_) {
    throw gcnew ObjectDisposedException("D3D9Backend");
  }
}

bool D3D9Backend::IsReadyForRendering::get() {
  return !isDisposed_ && isInitialized_ && !isFaulted_ && renderer_ != nullptr;
}

generic <typename TCapability> where TCapability : ref class
bool D3D9Backend::TryGetCapability(
    [System::Runtime::InteropServices::Out] TCapability % capability) {
  Type^ capabilityType = TCapability::typeid;
  Type^ instanceType = GetType();
  if (capabilityType == nullptr ||
      !capabilityType->IsAssignableFrom(instanceType)) {
    capability = TCapability();
    return false;
  }

  capability = safe_cast<TCapability>(this);
  return true;
}

}  // namespace FastDrawingVisual::Rendering::Backends
